### PR TITLE
implemented fisher yates shuffle algorithm

### PR DESCRIPTION
**Pull request by baykaner**
- this is useful in places where we want to shuffle values in vector deterministically with our lfg/lcg

### CODE REVIEW

File: libs/core/include/core/random.hpp [hunk 1]
```diff
@@ -27,5 +29,45 @@ struct Random
...
+ * @tparam T
+ */
+template <typename T>
+static void Shuffle(LaggedFibonacciGenerator<> & gen, std::vector<T> const &in_vec, std::vector<T> &out_vec)
+{
+  out_vec = in_vec;
+
+  for (std::size_t i = out_vec.size() - 1; i > 0; --i)
+  {
+    std::size_t j = gen() % (i + 1);
```
Comment (troels-im) [inline libs/core/include/core/random.hpp]: I would shift this by 19. LCG is known to be alternating on first bit and subsequent bits to be strongly correlated.
```suggestion
    std::size_t j = (gen() >> 19) % (i + 1);
```

File: libs/core/include/core/random.hpp [hunk 2]
```diff
@@ -27,5 +29,45 @@ struct Random
...
+ * @tparam T
+ */
+template <typename T>
+static void Shuffle(LinearCongruentialGenerator & gen, std::vector<T> const &in_vec, std::vector<T> &out_vec)
+{
+  out_vec = in_vec;
+
+  for (std::size_t i = out_vec.size() - 1; i > 0; --i)
+  {
+    std::size_t j = gen() % (i + 1);
```
Comment (troels-im) [inline libs/core/include/core/random.hpp]: Same here:
```suggestion
    std::size_t j = (gen()>>19) % (i + 1);
```

**Review by troels-im — CHANGES_REQUESTED** [review]
Some minor changes suggested, but otherwise it looks good.

File: libs/core/tests/random/shuffle_tests.cpp [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+//   limitations under the License.
+//
+//------------------------------------------------------------------------------
+
+#include "core/random.hpp"
+#include "vectorise/fixed_point/fixed_point.hpp"
+
+#include "gtest/gtest.h"
+
+#include <iostream>
```
Comment (wkaluza) [inline libs/core/tests/random/shuffle_tests.cpp]: Please do not include `<iostream>` unless it's actually used. It's a relatively large, expensive header to include, and it does our build times no favours.

**Review by troels-im — APPROVED** [review]
LGTM

**Pull request merged**